kernel/src/memory/bootstrap_heap.rs

```diff
@@ -0,0 +1,229 @@
+//! A simple bitmap allocator used to allocate memory for the buddy allocator
+
+#[cfg(test)]
+use std::boxed::Box;
+use core::{mem, ptr::{self, Unique}};
+use core::marker::PhantomData;
+use core::ops::{Deref, DerefMut};
+use spin::{Once, Mutex};
+use bit_field::BitField;
+use super::{physical_allocator::BLOCKS_IN_TREE, buddy_allocator::Block};
+
+/// How many 8 heap objects to allocate at maximum.
+// TODO two stage bootstrap -- allocate 1gib on the bootstrap heap and the rest on kernel heap
+const OBJECTS8_NUMBER: usize = 8; // 8 gives us 64GiB while taking 32mib
+
+pub static BOOTSTRAP_HEAP: BootstrapHeap = BootstrapHeap(Once::new());
+
+/// A holding struct for the bootstrap heap.
+pub struct BootstrapHeap(Once<BootstrapAllocator<[Block; BLOCKS_IN_TREE]>>);
+
+impl BootstrapHeap {
+    /// Allocates a zeroed object. Panics if bootstrap heap is not initialized
+    pub unsafe fn allocate_zeroed(&self) -> Option<BootstrapHeapBox<[Block; BLOCKS_IN_TREE]>> {
+        self.0.wait().unwrap().allocate_zeroed()
+    }
+
+    /// Initialises the bootstrap heap with a begin address.
+    pub unsafe fn init_unchecked(&self, address: usize) {
+        self.0.call_once(|| BootstrapAllocator::new_unchecked(address));
+    }
+
+    /// Get the end address of the bootstrap heap. Inclusive. Panics if uninitialized
+    pub fn end(&self) -> usize {
+        self.0.wait().unwrap().start() as usize +
+            BootstrapAllocator::<[Block; BLOCKS_IN_TREE]>::space_taken()
+    }
+}
+
+/// A bitmap heap/physmem allocator to bootstrap the buddy allocator since it requires a
+/// (relative to how much the stack should be used for) large amount of memory.
+#[derive(Debug)]
+pub struct BootstrapAllocator<T> {
+    start_addr: usize,
+    objects8: Mutex<[Objects8; OBJECTS8_NUMBER]>,
+    _phantom: PhantomData<T>,
+}
+
+/// A bitmap consisting of 8 objects
+#[derive(Debug, Copy, Clone)]
+struct Objects8(u8);
```

Comment:
Why not directly use a u8? The name Objects8 is also quite unclear as to what it is used for.

Reply:
I mean, the doc explains it, doesn't it? I would have called it 8Objects but that isn't allowed (begins with a number). I use a newtype for clarity.

Reply:
8Objects doesn't seem any more clear to me. What about rather abstracting access to the bitmap for clarity?

Reply:
Okay, I'll make it into just a plain u8.